common/models/templates/src/priority-queue.ts

```diff
@@ -4,216 +4,215 @@
-     * heap reconstruction.
-     * - O(`elements.count` * log(`heap.count`)) - logarithmic when elements.count << heap.count
-     * @param elements A group of elements to enqueue simultaneously.
-     */
-    enqueueAll(elements: Type[]) {
-      if(elements.length == 0) {
-        return;
-      }
-      
-      let firstIndex = this.count 
-      this.heap = this.heap.concat(elements);
-      let firstParent = PriorityQueue.parentIndex(firstIndex);
-
-      // The 'parent' of index 0 will return -1, which is illegal.
-      this.heapify(firstParent >= 0 ? firstParent : 0, PriorityQueue.parentIndex(this.count-1));
+  /**
+   * Efficiently batch-enqueues multiple elements.
+   * Worst-case is the _better_ of the following:
+   * - O(`elements.count` + `heap.count`) - large element counts will trigger in-place
+   * heap reconstruction.
+   * - O(`elements.count` * log(`heap.count`)) - logarithmic when elements.count << heap.count
+   * @param elements A group of elements to enqueue simultaneously.
+   */
+  enqueueAll(elements: Type[]) {
+    if(elements.length == 0) {
+      return;
     }
 
-    /**
-     * Removes the highest-priority element from the queue, returning it.
-     * <p>
-     * Is O(log N), where N = number of items in the priority queue.
-     */
-    dequeue(): Type {
-      if(this.count == 0) {
-        return undefined;
-      }
+    let firstIndex = this.count
+    this.heap = this.heap.concat(elements);
+    let firstParent = PriorityQueue.parentIndex(firstIndex);
 
-      const root = this.heap[0];
-      let tail = this.heap.pop();
-      if(this.heap.length > 0) {
-        this.heap[0] = tail;
-        this.siftDown(0);
-      }
+    // The 'parent' of index 0 will return -1, which is illegal.
+    this.heapify(firstParent >= 0 ? firstParent : 0, PriorityQueue.parentIndex(this.count-1));
+  }
 
-      return root;
+  /**
+   * Removes the highest-priority element from the queue, returning it.
+   * <p>
+   * Is O(log N), where N = number of items in the priority queue.
+   */
+  dequeue(): Type | undefined {
+    if(this.count == 0) {
+      return undefined;
     }
 
-    /**
-     * Compares the entry at the specified index against its children,
-     * propagating it downward within the heap until heap requirements are specified.
-     * <p>
-     * Is O(log N), where N = number of items in the priority queue.
-     * 
-     * @param index The index of the top-most node that must be examined
-     * for repositioning.
-     * @returns `true` if a swap occurred, `false` otherwise.
-     */
-    private siftDown(index: number): boolean {
-      let leftIndex = PriorityQueue.leftChildIndex(index);
-      let rightIndex = PriorityQueue.rightChildIndex(index);
-      let topMostIndex = index;
-
-      if(leftIndex < this.heap.length && this.comparator(this.heap[leftIndex], this.heap[topMostIndex]) < 0) {
-        topMostIndex = leftIndex;
-      }
+    const root = this.heap[0];
+    let tail = this.heap.pop() as Type;
```

Comment:
Why is this change required?

Reply:
Similar type-strictness shenanigans.  Note lines 158-160:  they logically prevent `.pop()` from returning `undefined`, but the TS engine isn't inferring that detail properly.